### PR TITLE
Introduce a common base type for SourceOrdinaryMethodSymbolBase and SourceUserDefinedOperatorSymbolBase to prepare for explicit implementations of operators

### PR DESCRIPTION
**Pull request by AlekseyTs**
Shared code is pulled from derived types to the new base. This is just a refactoring, no behavior changes intended.

However, there is a change in behavior that addresses #53069.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs:29]:
```suggestion

```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs:200]:
```suggestion
        }

        internal sealed override bool IsExplicitInterfaceImplementation
```

**Review by Youssef1313 — CHANGES_REQUESTED** [review]
nit: whitespace changes

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs:242]: Did you consider whether we should:
- pull this abstract member up to `SourceOrdinaryMethodOrUserDefinedOperatorSymbol` (and have its return type changed to match the signature in `SourceOrdinaryMethodSymbolBase`).
- remove the abstract member `SourceOrdinaryMethodSymbolBase.MakeParametersAndBindReturnType`
- pull usages up to `SourceOrdinaryMethodOrUserDefinedOperatorSymbol`

I don't feel it's important to make this change, but am hoping to get some insight into your process for deciding which members to extract or not by asking this.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs:242]: > Did you consider whether we should:
...

I intentionally left this method here and didn't try to unify it with method from SourceOrdinaryMethodSymbolBase because that one has to worry about things this this class doesn't have to worry about. At least at his moment I think that it doesn't have to worry about that.

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review, need a second review.

**Comment by AlekseyTs** [comment]
@RikkiGibson You might want to review new commits starting at the 5th one

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs:226]: Should this be sealed? It doesn't appear to be overridden by any derived types today.

**Review by 333fred — APPROVED** [review]
LGTM (commit 6). One minor comment.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodOrUserDefinedOperatorSymbol.cs:226]: > Should this be sealed? It doesn't appear to be overridden by any derived types today.

No, sealing a method is supposed to communicate intent to prevent derived types from changing implementation. I do not have this intent here.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**
